File: 2909-minimum-sum-of-mountain-triplets-ii/2909-minimum-sum-of-mountain-triplets-ii.py
class Solution:
    def minimumSum(self, nums: List[int]) -> int:
        s,n=0,len(nums)
        rm,mini=[],nums[-1]
        for i in range(n-1,-1,-1):
            rm.append(mini)
            if nums[i]<mini:
                mini=nums[i]
        mini,ans=nums[0],float('inf')
        for i in range(1,n-1):
            if nums[i]>mini and nums[i]>rm[n-1-i]:
                ans=min(ans,mini+nums[i]+rm[n-1-i])
            if nums[i]<mini:
                mini=nums[i]
        if ans==float('inf'):
            return -1
        return ans

File: 2909-minimum-sum-of-mountain-triplets-ii/2909-minimum-sum-of-mountain-triplets-ii.java
class Solution {
    public int minimumSum(int[] nums) {
        int n = nums.length;
        int[] prefixMin = new int[n];
        prefixMin[0] = nums[0];
        for(int i=1;i<n;i++){
            prefixMin[i] = Math.min(nums[i],prefixMin[i-1]);
        }
        
        int[] suffixMin = new int[n];
        suffixMin[n-1] = nums[n-1];
        for(int i=n-2;i>=0;i--){
            suffixMin[i] = Math.min(nums[i],suffixMin[i+1]);
        }
        int ans = Integer.MAX_VALUE;
        boolean found = false;
        for(int i=0;i<n;i++){
            if(prefixMin[i]<nums[i] && suffixMin[i]<nums[i]){
                found = true;
                ans = Math.min(prefixMin[i]+nums[i]+suffixMin[i],ans);
            }
        }
        if(found){
           return ans ;
        }else{
            return -1;
        }
        
    }
}